compiler/rustc_ast/src/node_id.rs

```diff
@@ -2,19 +2,25 @@ use rustc_span::ExpnId;
 use std::fmt;
 
 rustc_index::newtype_index! {
+    /// Identifies an AST node.
+    ///
+    /// This identifies top-level definitions, expressions, and everything in between.
+    /// This is later turned into [`DefId`] and `HirId` for the HIR.
+    ///
+    /// [`DefId`]: rustc_span::def_id::DefId
     pub struct NodeId {
         DEBUG_FORMAT = "NodeId({})"
     }
 }
 
 rustc_data_structures::define_id_collections!(NodeMap, NodeSet, NodeId);
 
-/// `NodeId` used to represent the root of the crate.
+/// The [`NodeId`] used to represent the root of the crate.
 pub const CRATE_NODE_ID: NodeId = NodeId::from_u32(0);
 
 /// When parsing and doing expansions, we initially give all AST nodes this AST
-/// node value. Then later, during expansion, we renumber them to have small,
-/// positive ids.
+/// [`NodeId`]. Then later, during expansion, we renumber them to have small,
```

Comment:
This part is confusing to me: it says that "when parsing and doing expansions" we use a dummy ID, but "then later, during expansion" we create real IDs. Does it mean that at the beginning of expansions we use dummy IDs but a later phase of it we create real ones?

Reply:
That's correct - the creation of real `NodeId`s is done here: https://github.com/rust-lang/rust/blob/9bc8b00b4a4e38ccbc3aeec2c123538973c67eba/compiler/rustc_expand/src/expand.rs#L1702-L1707